# Problem 19

You are given the following information, but you may prefer to do some research for yourself.

1 Jan 1900 was a Monday.

Thirty days has September,

April, June and November.

All the rest have thirty-one,

Saving February alone,

Which has twenty-eight, rain or shine.

And on leap years, twenty-nine.

A leap year occurs on any year evenly divisible by 4, but not on a century unless it is divisible by 400.

How many Sundays fell on the first of the month during the twentieth century (1 Jan 1901 to 31 Dec 2000)?

In [ ]:
import datetime as dt
from itertools import cycle
import pandas as pd

In [81]:
normalyear = {1 : 31, 2 : 28, 3 : 31,
         4 : 30, 5 : 31, 6 : 30,
         7 : 31, 8 : 31, 9 : 30,
         10 : 31, 11 : 30, 12 : 31}

leapyear = {1 : 31, 2 : 29, 3 : 31,
         4 : 30, 5 : 31, 6 : 30,
         7 : 31, 8 : 31, 9 : 30,
         10 : 31, 11 : 30, 12 : 31}

daysofweek = cycle(["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"])

In [82]:
dates = []
for year in range(1900, 2010):
    #Check if year is leap and choose the appropriate calendar
    if (year % 4 == 0 and year % 100 != 0) or year % 400 == 0:
        months = leapyear
    else:
        months = normalyear
        
    for month in months.keys():
        for day in range(1, months[month]+1):
            dates.append( [year, month, day, next(daysofweek)] )

In [83]:
df = pd.DataFrame(dates, columns = ["Year", "Month", "Day", "Day_of_week"])

In [84]:
filt = (df["Year"] > 1900) & (df["Year"] < 2001) & (df["Day"] == 1) & (df["Day_of_week"] == "Sun")
df.loc[filt]

Year  Month  Day Day_of_week
608    1901      9    1         Sun
699    1901     12    1         Sun
881    1902      6    1         Sun
1126   1903      2    1         Sun
1154   1903      3    1         Sun
...     ...    ...  ...         ...
35825  1998      2    1         Sun
35853  1998      3    1         Sun
36098  1998     11    1         Sun
36371  1999      8    1         Sun
36798  2000     10    1         Sun

[171 rows x 4 columns]

# Answer

Answer is how many rows remained in data frame after we had used filter (171).